### PR TITLE
MAINT: Enable type checking for subset of shap.plots, update to PEP604

### PR DESCRIPTION
**Pull request by connortann**
## Overview

Supports #3730

Gets mypy passing on about half the modules in shap.plots with `check-untyped-defs=True`.

## Discussion

Also updates types hints to use the nice PEP604-style annotations:
- `X | Y` rather than `Union[X, Y]`
- `X | None` rather than `Optional[X]`.
 
This is made possible even with python 3.9 support with `__future__` imports.

As a supporting change, enables two typing-related linting rules.

### CODE REVIEW

File: shap/_explanation.py [hunk 1]
```diff
@@ -170,7 +172,7 @@ def __init__(
         )
 
     @property
-    def shape(self) -> tuple[Union[int, None], ...]:
+    def shape(self) -> tuple[int | None, ...]:
```
Comment (CloseChoice) [inline shap/_explanation.py:175]: this won't work well on python3.9. I think we could drop support for python3.9 altogether though.

**Review by CloseChoice — CHANGES_REQUESTED** [review]
looks great just one comment we should sort out

File: shap/_explanation.py (hunk 1 above)
Comment (connortann) [reply shap/_explanation.py:175]: My understanding is that this annotation works perfect fine on python 3.9 due to the [future](https://docs.python.org/3/library/__future__.html) import, specifically:

```python
from __future__ import annotations
```

I first noticed that this was possible on the streamlit project (e.g. [here](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/runtime/fragment.py) ), which supports python 3.8+ and yet uses the `X | Y` syntax for annotations.

In this PR the tests still pass on python 3.9, so there don't appear to be any runtime issues.

> this won't work well on python3.9.

Are there any other aspects or downsides that we should consider?

Comment (connortann) [reply shap/_explanation.py:175]: On a related note, I found two neat ruff linting rules which we can enable for consistent use of type hints:
- [FA](https://docs.astral.sh/ruff/rules/#flake8-future-annotations-fa): future annotations
- [TCH](https://docs.astral.sh/ruff/rules/#flake8-type-checking-tch): type checking

I pushed them to this branch, which required a few small modifications to get them passing.

**Review by CloseChoice — APPROVED** [review]
looks good

**Pull request merged**